parsec/mca/device/cuda/device_cuda_module.c

```diff
@@ -1295,7 +1306,7 @@ parsec_gpu_data_stage_in( parsec_device_cuda_module_t* cuda_device,
     transfer_from = parsec_data_start_transfer_ownership_to_copy(original, gpu_device->super.device_index, (uint8_t)type);
 
     if( PARSEC_FLOW_ACCESS_WRITE & type && gpu_task->task_type != PARSEC_GPU_TASK_TYPE_PREFETCH ) {
-        gpu_elem->version++;  /* on to the next version */
+        gpu_elem->version = candidate->version + 1;  /* on to the next version */
```

Comment:
I agree we need to do this here, but I think we also need to fix at line 1487 (before the patch) / 1506 (after the patch): we do `gpu_elem->version = candidate->version;` when the transfer (if a transfer is needed) is done, and the version of `candidate` is one less than what we want, and this will overwrite the effect of this action here.

This doesn't typically fail with GEMM or POTRF as benchmarks / tests because they only send the result back to the host memory once, and then we discard / release the GPU copy, so the fact that the version is wrong doesn't trigger a problem, but if you remove the release_memory from the test this should create a problem.

Reply:
My problem is I don't understand (neither in the code before the PR or inside the PR) what is supposed to increase the version of the GPU copy to what I believe it should be.

I did a run (or the PR, at commit fc82786a2c791ec65c2bc3f0b01d1086229b3fbf) with M=1 tile, N=1 tile, K=2 tiles, a single GPU under gdb.

I break on `parsec_cuda_kernel_scheduler` to see all kernels passed to the GPU, and on `parsec_gpu_data_reserve_device_space`. When I'm in kernel scheduler, I print the task name. When I'm in reserve_device_space, I go until the GPU copy for C is allocated (line 927), then I install a watchpoint on any change of this copy's version.

Here is what I see:
  - we start GEMM(0, 0, 0), we allocate the GPU copy for C and set the version to 0 (inside the allocation)
  - version goes to UINT_MAX at line 1149 according to the patch
  - it is then set to 2 at line 1311, because it's a write (C) and we take candidate (CPU copy)'s version + 1. transfer_from is 0, everything is logic
  - next step, it is reset to 1 at line 1512 after we have started the copy. You seem to expect that, OK, why not.
  - then it remains unmodified until we start scheduling GEMM(0, 0, 1) on the GPU. But definitely here, its version should be 2. Shouldn't it? The CPU version is 2 (has been updated by the DSL at the completion of the task), so it appears like if the GPU copy is too old (which it isn't).
  - then we reach the call to `parsec_data_start_transfer_ownership_to_copy` at line 1308:
     - original->owner_device is set to 2, and the GPU copy is in the state SHARED.
     - there is no other copy with the state OWNED, and all other copies are in the state SHARED, so we never check if another copy has a version higher than the GPU version.
     - that explains why we don't consider a new transfer is necessary, and we return -1
  - then we set the GPU copy version to 3 directly, by line 1311 (candidate->version+1)
  - we don't schedule the transfer, so we don't reset the version to candidate->version

It works, but I'm surprised that this is the expected behavior. I would have expected the GPU copy to get the version 2 when the transfer is complete, and the transfer_ownership to return -1 because the GPU copy has the highest version. At the time transfer_ownership returns -1, it looks to me it returns -1 for the wrong reason.

Reply:
This is not what I expect to see here. But, that is another problem for another PR, here we are just fixing the race condition between accelerator copies management.